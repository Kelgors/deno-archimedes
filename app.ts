import { Hono } from 'hono';
import { trimTrailingSlash } from 'hono/trailing-slash';
import { Client } from 'postgres';
import routers from './routers/main.ts';
import { HonoEnv } from './types/hono.ts';

export function createApp(db: Client) {
  const app = new Hono<HonoEnv>();

  app.use(trimTrailingSlash());

  app.use(async (c, next) => {
    c.set('db', db);
    await next();
  });

  app.route('/', routers);

  return app;
}
