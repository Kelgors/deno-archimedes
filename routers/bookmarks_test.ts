import { expect } from '@std/expect';
import { after, before, describe, it } from '@std/testing/bdd';
import { assertSnapshot } from '@std/testing/snapshot';
import { createApp } from '../app.ts';
import { createClient } from '../db/mod.ts';

const db = createClient();
const app = createApp(db);
before(async () => {
  await db.connect();
  const populateScript = await Deno.readTextFile('db/seeds/01-bookmarks.sql');
  await db.queryArray(populateScript);
});
after(() => db.end());

let createItemId: string;

describe('GET /bookmarks', () => {
  it('should return bookmarks', async (t) => {
    const result = await app.request('/api/bookmarks');
    expect(result.status).toBe(200);
    assertSnapshot(t, await result.json());
  });
});

describe('GET /bookmarks/:id', () => {
  it('should return google bookmark ', async (t) => {
    const result = await app.request('/api/bookmarks/99943bac-567a-4bee-ba4d-fc72fed4c26b');
    expect(result.status).toBe(200);
    assertSnapshot(t, await result.json());
  });

  it('should return a bad request when id param is not a uuid', async () => {
    const result = await app.request('/api/bookmarks/1');
    expect(result.status).toBe(400);
  });

  it('should return a not found response', async () => {
    const result = await app.request('/api/bookmarks/99999999-9999-4999-9999-999999999999');
    expect(result.status).toBe(404);
  });
});

describe('POST /bookmarks', () => {
  it('should create a bookmark', async (t) => {
    const result = await app.request('/api/bookmarks', {
      method: 'POST',
      body: JSON.stringify({
        name: 'name',
        url: 'http://url',
        description: 'description\n',
      }),
      headers: { 'Content-Type': 'application/json' },
    });
    expect(result.status).toBe(201);
    const body = await result.json();
    assertSnapshot(t, { ...body.data, id: undefined });
    expect(body.data.id).toHaveLength(36);
    createItemId = body.data.id;
  });

  it('should return a bad request response when body is not valid', async () => {
    const result = await app.request('/api/bookmarks', { method: 'POST' });
    expect(result.status).toBe(400);
  });
});

describe('PUT /bookmarks/:id', () => {
  it('should update a bookmark', async (t) => {
    const result = await app.request(`/api/bookmarks/${createItemId}`, {
      method: 'PUT',
      body: JSON.stringify({
        name: 'updated-name',
        url: 'http://updated-url',
        description: 'updated description\n',
      }),
      headers: { 'Content-Type': 'application/json' },
    });
    expect(result.status).toBe(200);
    const body = await result.json();
    assertSnapshot(t, { ...body.data, id: undefined });
    expect(body.data.id).toBe(createItemId);
  });

  it('should return a bad request when id param is not a uuid', async () => {
    const result = await app.request('/api/bookmarks/1');
    expect(result.status).toBe(400);
  });

  it('should return a not found response', async () => {
    const result = await app.request('/api/bookmarks/99999999-9999-4999-9999-999999999999');
    expect(result.status).toBe(404);
  });

  it('should return a bad request response when body is not valid', async () => {
    const result = await app.request('/api/bookmarks/99999999-9999-4999-9999-999999999999', { method: 'PUT' });
    expect(result.status).toBe(400);
  });
});

describe('DELETE /bookmarks/:id', () => {
  it('should delete a bookmark', async () => {
    const result = await app.request(`/api/bookmarks/${createItemId}`, { method: 'DELETE' });
    expect(result.status).toBe(200);
    await expect(result.json()).resolves.toEqual({ success: true });
  });

  it('should return a bad request when id param is not a uuid', async () => {
    const result = await app.request('/api/bookmarks/1');
    expect(result.status).toBe(400);
  });

  it('should return a not found response', async () => {
    const result = await app.request('/api/bookmarks/99999999-9999-4999-9999-999999999999');
    expect(result.status).toBe(404);
  });
});
