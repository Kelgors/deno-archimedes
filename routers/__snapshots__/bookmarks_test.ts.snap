export const snapshot = {};

snapshot[`global > GET /bookmarks > should return bookmarks 1`] = `
{
  data: [
    {
      description: "The Internet privacy company that empowers you to seamlessly take control of your personal information online, without any tradeoffs.",
      id: "3a543337-0989-400b-9d39-c2f02270a299",
      name: "DuckDuckGo",
      url: "https://duckduckgo.com/",
    },
    {
      description: "The search engine that values you as a user, not as a product",
      id: "7abd6fbd-dc99-40b8-a404-efe6ae8a51f0",
      name: "Qwant",
      url: "https://www.qwant.com/",
    },
    {
      description: "",
      id: "99943bac-567a-4bee-ba4d-fc72fed4c26b",
      name: "Google",
      url: "https://www.google.com/",
    },
    {
      description: "",
      id: "96bd0559-5a64-4d57-b702-c9b7f6c6ba3d",
      name: "Bing",
      url: "https://www.bing.com/",
    },
  ],
}
`;

snapshot[`global > GET /bookmarks/:id > should return google bookmark  1`] = `
{
  data: {
    description: "",
    id: "99943bac-567a-4bee-ba4d-fc72fed4c26b",
    name: "Google",
    url: "https://www.google.com/",
  },
}
`;

snapshot[`global > POST /bookmarks > should create a bookmark 1`] = `
{
  description: "description
",
  id: undefined,
  name: "name",
  url: "http://url",
}
`;

snapshot[`global > PUT /bookmarks/:id > should update a bookmark 1`] = `
{
  description: "updated description
",
  id: undefined,
  name: "updated-name",
  url: "http://updated-url",
}
`;
